support bloom filters that have more `include` items than `exclude`

If `len(include) > len(exclude)` then then an error is thrown:

```
/deps/lib/python3.6/site-packages/filtercascade/__init__.py:268: in cascade_with_characteristics
    [Bloomer.filter_with_characteristics(capacity, error_rates[0])],
/deps/lib/python3.6/site-packages/filtercascade/__init__.py:75: in filter_with_characteristics
    return Bloomer(size=size, nHashFuncs=nHashFuncs, level=level)
/deps/lib/python3.6/site-packages/filtercascade/__init__.py:30: in __init__
    self.bitarray = bitarray.bitarray(self.size, endian='little')
E   ValueError: cannot create bitarray with negative length
```

For the addon blocklist it's possible that we will have at some point more blocked addons (include) than not blocked.

--- a/filtercascade/__init__.py
+++ b/filtercascade/__init__.py
@@ -22,14 +22,15 @@ class HashAlgorithm(IntEnum):
 
 
 class InvertedLogicException(Exception):
-    def __init__(self, *, exclude_count, include_len):
+    def __init__(self, *, depth, exclude_count, include_len):
         self.message = (
-            "Inefficient filter. Exclude count was smaller than the include "
-            f"count ({exclude_count}<{include_len}). If you reached this "
-            "exception, then your input data had an uncountable iterator and "
-            "the only way to fix this issue is via a code update where you "
-            "manually swap the input 'include' and 'exclude' parameters and "
-            "set the invertedLogic boolean to True upon construction."
+            f"At Depth {depth}, exclude set ({exclude_count}) was < include set "
+            f"({include_len}). If you reached this exception, then either your "
+            "input data had an uncountable iterator, or your filter length was "
+            "insufficient, and you need to increase it. The only way to fix this "
+            "issue is via a code update where you either manually swap the input "
+            "'include' and  'exclude' parameters and set the invertedLogic boolean "
+            "to True upon construction, or set a larger min_filter_length."
         )
 
     def __str__(self):
@@ -142,7 +143,9 @@ class Bloomer:
     @classmethod
     def calc_size(cls, nHashFuncs, elements, falsePositiveRate):
         # From CRLite paper, https://cbw.sh/static/pdf/larisch-oakland17.pdf
-        return math.ceil(1.44 * elements * math.log(1 / falsePositiveRate, 2))
+        min_bits = math.ceil(1.44 * elements * math.log(1 / falsePositiveRate, 2))
+        # Ensure the result is divisible by 8 for full bytes
+        return 8 * math.ceil(min_bits / 8)
 
     @classmethod
     def from_buf(cls, buf, salt=None):
@@ -150,23 +153,27 @@ class Bloomer:
         hashAlgInt, size, nHashFuncs, level = Bloomer.layer_struct.unpack(
             buf[: Bloomer.layer_struct.size]
         )
+        buf = buf[Bloomer.layer_struct.size :]
+
         byte_count = math.ceil(size / 8)
         ba = bitarray.bitarray(endian="little")
-        ba.frombytes(buf[10 : 10 + byte_count])
+
+        ba.frombytes(buf[:byte_count])
+        buf = buf[byte_count:]
+
         bloomer = Bloomer(
-            size=1,
+            size=size,
             nHashFuncs=nHashFuncs,
             level=level,
             hashAlg=HashAlgorithm(hashAlgInt),
             salt=salt,
         )
-        bloomer.size = size
         log.debug(
-            "Size is {}, level {}, nHashFuncs, {}".format(size, level, nHashFuncs)
+            f"from_buf size is {size} ({byte_count} bytes), level {level}, nHashFuncs, {nHashFuncs}"
         )
         bloomer.bitarray = ba
 
-        return (buf[10 + byte_count :], bloomer)
+        return (buf, bloomer)
 
 
 class FilterCascade:
@@ -312,7 +319,7 @@ class FilterCascade:
 
             if exclude_count < include_len:
                 raise InvertedLogicException(
-                    exclude_count=exclude_count, include_len=include_len
+                    depth=depth, exclude_count=exclude_count, include_len=include_len
                 )
 
             endtime = datetime.datetime.utcnow()
@@ -450,6 +457,7 @@ class FilterCascade:
         while len(buf) > 0:
             (buf, f) = Bloomer.from_buf(buf)
             filters.append(f)
+        assert len(buf) == 0, "buffer should be consumed"
 
         return FilterCascade(
             filters, version=version, hashAlg=hashAlg, salt=salt, invertedLogic=inverted

--- a/filtercascade/__init__.py
+++ b/filtercascade/__init__.py
@@ -21,6 +21,21 @@ class HashAlgorithm(IntEnum):
     SHA256 = 2
 
 
+class InvertedLogicException(Exception):
+    def __init__(self, *, exclude_count, include_len):
+        self.message = (
+            "Inefficient filter. Exclude count was smaller than the include "
+            f"count ({exclude_count}<{include_len}). If you reached this "
+            "exception, then your input data had an uncountable iterator and "
+            "the only way to fix this issue is via a code update where you "
+            "manually swap the input 'include' and 'exclude' parameters and "
+            "set the invertedLogic boolean to True upon construction."
+        )
+
+    def __str__(self):
+        return self.message
+
+
 # A simple-as-possible bloom filter implementation making use of version 3 of the 32-bit murmur
 # hash function (for compat with multi-level-bloom-filter-js).
 # mgoodwin 2018
@@ -186,7 +201,14 @@ class FilterCascade:
         version=2,
         hashAlg=HashAlgorithm.MURMUR3,
         salt=None,
+        invertedLogic=None,
     ):
+        """
+            Construct a FilterCascade.
+            error_rates: If not supplied, defaults will be calculated
+            invertedLogic: If not supplied (or left as None), it will be auto-
+                detected.
+        """
         self.filters = filters
         self.error_rates = error_rates
         self.growth_factor = growth_factor
@@ -194,6 +216,7 @@ class FilterCascade:
         self.version = version
         self.hashAlg = hashAlg
         self.salt = salt
+        self.invertedLogic = invertedLogic
 
         if self.salt and version < 2:
             raise ValueError("salt requires format version 2 or greater")
@@ -221,6 +244,15 @@ class FilterCascade:
             len(include)
         except TypeError as te:
             raise TypeError("include is not a list", te)
+
+        if self.invertedLogic is None:
+            try:
+                self.invertedLogic = len(include) > len(exclude)
+            except TypeError:
+                self.invertedLogic = False
+
+        if self.invertedLogic:
+            include, exclude = exclude, include
 
         include_len = len(include)
 
@@ -272,9 +304,16 @@ class FilterCascade:
             # that *includes* the false positives and *excludes* the true positives
             log.debug("Processing false positives")
             false_positives = set()
+            exclude_count = 0
             for elem in exclude:
+                exclude_count += 1
                 if elem in filter:
                     false_positives.add(elem)
+
+            if exclude_count < include_len:
+                raise InvertedLogicException(
+                    exclude_count=exclude_count, include_len=include_len
+                )
 
             endtime = datetime.datetime.utcnow()
             log.debug(
@@ -319,7 +358,7 @@ class FilterCascade:
         if depth < len(self.filters):
             del self.filters[depth:]
 
-    def __contains__(self, elem):
+    def __internal_contains__(self, elem):
         for layer, filter in [
             (idx + 1, self.filters[idx]) for idx in range(len(self.filters))
         ]:
@@ -329,6 +368,10 @@ class FilterCascade:
                     return even is not True
             else:
                 return even is not False
+
+    def __contains__(self, elem):
+        result = self.__internal_contains__(elem)
+        return not result if self.invertedLogic is True else result
 
     @deprecated(
         version="0.2.3",
@@ -369,7 +412,9 @@ class FilterCascade:
         if self.version > 1:
             salt_len = len(self.salt) if self.salt else 0
             f.write(
-                FilterCascade.version_2_salt_struct.pack(False, self.hashAlg, salt_len)
+                FilterCascade.version_2_salt_struct.pack(
+                    self.invertedLogic, self.hashAlg, salt_len
+                )
             )
             if self.salt:
                 f.write(self.salt)
@@ -406,7 +451,9 @@ class FilterCascade:
             (buf, f) = Bloomer.from_buf(buf)
             filters.append(f)
 
-        return FilterCascade(filters, version=version, hashAlg=hashAlg, salt=salt)
+        return FilterCascade(
+            filters, version=version, hashAlg=hashAlg, salt=salt, invertedLogic=inverted
+        )
 
     @classmethod
     def loadDiffMeta(cls, f):

--- a/filtercascade/test.py
+++ b/filtercascade/test.py
@@ -43,12 +43,12 @@ def predictable_serial_gen(end):
         yield m.hexdigest()
 
 
-def get_serial_sets(*, num_revoked, num_valid):
-    valid = predictable_serial_gen(num_revoked + num_valid)
+def get_serial_iterator_and_set(*, num_set, num_iterator):
+    iterator_data = predictable_serial_gen(num_iterator + num_set)
     # revocations must be disjoint from the main set, so
     # slice off a set and re-use the remainder
-    revoked = set(islice(valid, num_revoked))
-    return (valid, revoked)
+    set_data = set(islice(iterator_data, num_set))
+    return (iterator_data, set_data)
 
 
 class TestFilterCascade(unittest.TestCase):
@@ -64,6 +64,7 @@ class TestFilterCascade(unittest.TestCase):
         for i in range(0, len(f1.filters)):
             self.assertBloomerEqual(f1.filters[i], f2.filters[i])
         self.assertEqual(f1.salt, f2.salt)
+        self.assertEqual(f1.hashAlg, f2.hashAlg)
 
     def test_bloomer_serial_deserial(self):
         b1 = filtercascade.Bloomer(size=32, nHashFuncs=6, level=1)
@@ -76,7 +77,7 @@ class TestFilterCascade(unittest.TestCase):
 
     def test_fc_serial_deserial(self):
         f1 = filtercascade.FilterCascade([])
-        f1.initialize(include=["A", "B", "C"], exclude=["D"])
+        f1.initialize(exclude=["A", "B", "C"], include=["D"])
 
         h = MockFile()
         f1.tofile(h)
@@ -117,8 +118,11 @@ class TestFilterCascade(unittest.TestCase):
     def test_fc_iterable(self):
         f = filtercascade.FilterCascade([])
 
-        valid, revoked = get_serial_sets(num_valid=500_000, num_revoked=3_000)
-        f.initialize(include=revoked, exclude=valid)
+        iterator, small_set = get_serial_iterator_and_set(
+            num_iterator=500_000, num_set=3_000
+        )
+        f.initialize(include=small_set, exclude=iterator)
+        self.assertFalse(f.invertedLogic)
 
         self.assertEqual(len(f.filters), 3)
         self.assertEqual(f.filters[0].size, 81272)
@@ -133,17 +137,19 @@ class TestFilterCascade(unittest.TestCase):
         """
         fc = filtercascade.FilterCascade([])
 
-        valid, revoked = get_serial_sets(num_valid=10, num_revoked=1)
-        fc.initialize(include=revoked, exclude=valid)
+        iterator, small_set = get_serial_iterator_and_set(num_iterator=10, num_set=1)
+        fc.initialize(include=small_set, exclude=iterator)
 
         with self.assertRaises(AssertionError):
-            valid2, revoked2 = get_serial_sets(num_valid=10, num_revoked=2)
-            fc.verify(include=revoked2, exclude=valid2)
+            iterator2, small_set2 = get_serial_iterator_and_set(
+                num_iterator=10, num_set=2
+            )
+            fc.verify(include=small_set2, exclude=iterator2)
 
     def test_fc_load_version_1(self):
         fc = filtercascade.FilterCascade([], version=1)
-        valid, revoked = get_serial_sets(num_valid=10, num_revoked=1)
-        fc.initialize(include=revoked, exclude=valid)
+        iterator, small_set = get_serial_iterator_and_set(num_iterator=10, num_set=1)
+        fc.initialize(include=small_set, exclude=iterator)
 
         h = MockFile()
         fc.tofile(h)
@@ -153,8 +159,8 @@ class TestFilterCascade(unittest.TestCase):
 
     def test_fc_load_version_2(self):
         fc = filtercascade.FilterCascade([], version=2)
-        valid, revoked = get_serial_sets(num_valid=10, num_revoked=1)
-        fc.initialize(include=revoked, exclude=valid)
+        iterator, small_set = get_serial_iterator_and_set(num_iterator=10, num_set=1)
+        fc.initialize(include=small_set, exclude=iterator)
 
         h = MockFile()
         fc.tofile(h)
@@ -166,14 +172,61 @@ class TestFilterCascade(unittest.TestCase):
         fc = filtercascade.FilterCascade(
             [], version=2, salt=b"nacl", hashAlg=filtercascade.HashAlgorithm.SHA256
         )
-        valid, revoked = get_serial_sets(num_valid=10, num_revoked=1)
-        fc.initialize(include=revoked, exclude=valid)
+        iterator, small_set = get_serial_iterator_and_set(num_iterator=10, num_set=1)
+        fc.initialize(include=small_set, exclude=iterator)
+        self.assertFalse(fc.invertedLogic)
 
         h = MockFile()
         fc.tofile(h)
 
         fc2 = filtercascade.FilterCascade.from_buf(h)
         self.assertFilterCascadeEqual(fc, fc2)
+
+    def test_fc_inverted_logic_iterators(self):
+        fc = filtercascade.FilterCascade([])
+        self.assertFalse(fc.invertedLogic)
+
+        iterator, huge_set = get_serial_iterator_and_set(
+            num_iterator=100, num_set=50_000
+        )
+        with self.assertRaises(filtercascade.InvertedLogicException):
+            fc.initialize(include=huge_set, exclude=iterator)
+
+    def test_fc_inverted_logic_automatic(self):
+        fc = filtercascade.FilterCascade([])
+        self.assertEqual(None, fc.invertedLogic)
+
+        iterator, huge_set = get_serial_iterator_and_set(
+            num_iterator=100, num_set=50_000
+        )
+
+        # Should automatically invert the logic
+        fc.initialize(include=huge_set, exclude=set(iterator))
+        self.assertTrue(fc.invertedLogic)
+
+        iterator, huge_set = get_serial_iterator_and_set(
+            num_iterator=100, num_set=50_000
+        )
+        fc.verify(include=huge_set, exclude=iterator)
+
+        h = MockFile()
+        fc.tofile(h)
+        fc2 = filtercascade.FilterCascade.from_buf(h)
+        self.assertFilterCascadeEqual(fc, fc2)
+
+        iterator, huge_set = get_serial_iterator_and_set(
+            num_iterator=100, num_set=50_000
+        )
+        fc2.verify(include=huge_set, exclude=iterator)
+
+    def test_fc_inverted_logic_explicit(self):
+        fc = filtercascade.FilterCascade([], invertedLogic=True)
+        iterator, small_set = get_serial_iterator_and_set(num_iterator=2, num_set=2)
+        fc.initialize(include=small_set, exclude=set(iterator))
+        self.assertTrue(fc.invertedLogic)
+
+        iterator, small_set = get_serial_iterator_and_set(num_iterator=2, num_set=2)
+        fc.verify(include=small_set, exclude=iterator)
 
 
 class TestFilterCascadeSalts(unittest.TestCase):
@@ -194,8 +247,8 @@ class TestFilterCascadeSalts(unittest.TestCase):
             [], hashAlg=filtercascade.HashAlgorithm.SHA256, salt=b"happiness"
         )
 
-        valid, revoked = get_serial_sets(num_valid=10, num_revoked=1)
-        fc.initialize(include=revoked, exclude=valid)
+        iterator, small_set = get_serial_iterator_and_set(num_iterator=10, num_set=1)
+        fc.initialize(include=small_set, exclude=iterator)
 
         self.assertEqual(len(fc.filters), 1)
         self.assertEqual(fc.bitCount(), 81272)
@@ -218,8 +271,8 @@ class TestFilterCascadeAlgorithms(unittest.TestCase):
     def verify_minimum_sets(self, *, hashAlg):
         fc = filtercascade.FilterCascade([], hashAlg=hashAlg)
 
-        valid, revoked = get_serial_sets(num_valid=10, num_revoked=1)
-        fc.initialize(include=revoked, exclude=valid)
+        iterator, small_set = get_serial_iterator_and_set(num_iterator=10, num_set=1)
+        fc.initialize(include=small_set, exclude=iterator)
 
         self.assertEqual(len(fc.filters), 1)
         self.assertEqual(fc.bitCount(), 81272)
@@ -229,8 +282,8 @@ class TestFilterCascadeAlgorithms(unittest.TestCase):
         self.assertEqual(len(f.data), 10174)
 
         fc2 = filtercascade.FilterCascade.from_buf(f)
-        valid2, revoked2 = get_serial_sets(num_valid=10, num_revoked=1)
-        fc2.verify(include=revoked2, exclude=valid2)
+        iterator2, small_set2 = get_serial_iterator_and_set(num_iterator=10, num_set=1)
+        fc2.verify(include=small_set2, exclude=iterator2)
 
     def test_murmurhash3(self):
         self.verify_minimum_sets(hashAlg=filtercascade.HashAlgorithm.MURMUR3)

--- a/filtercascade/test.py
+++ b/filtercascade/test.py
@@ -57,6 +57,7 @@ class TestFilterCascade(unittest.TestCase):
         self.assertEqual(b1.size, b2.size)
         self.assertEqual(b1.level, b2.level)
         self.assertEqual(b1.hashAlg, b2.hashAlg)
+        self.assertEqual(b1.bitarray.length(), b2.bitarray.length())
         self.assertEqual(b1.bitarray, b2.bitarray)
 
     def assertFilterCascadeEqual(self, f1, f2):
@@ -182,6 +183,21 @@ class TestFilterCascade(unittest.TestCase):
         fc2 = filtercascade.FilterCascade.from_buf(h)
         self.assertFilterCascadeEqual(fc, fc2)
 
+    def test_fc_small_filter_length(self):
+        fc = filtercascade.FilterCascade([], min_filter_length=8)
+
+        iterator, small_set = get_serial_iterator_and_set(
+            num_iterator=5_000, num_set=100
+        )
+
+        fc.initialize(include=small_set, exclude=iterator)
+        h = MockFile()
+        fc.tofile(h)
+        self.assertEqual(len(h.data), 281)
+
+        fc2 = filtercascade.FilterCascade.from_buf(h)
+        self.assertFilterCascadeEqual(fc, fc2)
+
     def test_fc_inverted_logic_iterators(self):
         fc = filtercascade.FilterCascade([])
         self.assertFalse(fc.invertedLogic)
@@ -193,7 +209,7 @@ class TestFilterCascade(unittest.TestCase):
             fc.initialize(include=huge_set, exclude=iterator)
 
     def test_fc_inverted_logic_automatic(self):
-        fc = filtercascade.FilterCascade([])
+        fc = filtercascade.FilterCascade([], min_filter_length=1024)
         self.assertEqual(None, fc.invertedLogic)
 
         iterator, huge_set = get_serial_iterator_and_set(
@@ -211,6 +227,9 @@ class TestFilterCascade(unittest.TestCase):
 
         h = MockFile()
         fc.tofile(h)
+
+        self.assertEqual(len(h.data), 1056)
+
         fc2 = filtercascade.FilterCascade.from_buf(h)
         self.assertFilterCascadeEqual(fc, fc2)
 

